feat(bar): send commands by mouse/touchpad/screen

CtByte opened the pull request:
This commit makes it possible to send commands from the bar by using the
mouse/touchpad/touchscreen.

Komorebi or custom commands can be sent by clicking on the
mouse's primary, secondary, middle, back or forward buttons.

As the primary single click is already used by widgets, only primary double clicks can
send commands. This limitation is due to Egui also triggering 2 single clicks
before a double click is triggered. Egui does not have an
implementation for stopping event propagation out of the box and would
be too much work to include.

Similarly, commands can be sent on every "tick" of mouse scrolling, touchpad or
touchscreen swiping in any of the 4 directions. This "tick" can be
adjusted to fit user's preference.

This is due to the fact, that Egui does not have an event for when a mouse "tick" occurs. It instead
gives a number of points that the user scrolled/swiped on each frame.

```json
  "mouse": {
    // The variable will be resolved
    //"on_primary_double_click": "komorebic replace-configuration $Env:USERPROFILE/komorebi.json",
    
    // This is the same as below but executed in a persistent powershell instance
    // "on_primary_double_click": "komorebic toggle-pause",
    "on_primary_double_click": {
      "focus_monitor_at_cursor": false,
      "ignore_mouse_follows_focus": false,
      "message": {
        "type": "TogglePause"
      }
    },
    "on_secondary_click": {
      "message": {
        "type": "NewWorkspace"
      }
    },
    "on_middle_click": {
      "message": {
        "type": "CloseWorkspace"
      }
    },
    "on_extra2_click": {
      "message": {
        "type": "CycleStack",
        "content": "Next"
      }
    },
    "on_extra1_click": {
      "message": {
        "type": "CycleStack",
        "content": "Previous"
      }
    },
    "vertical_scroll_threshold": 100,
    "on_scroll_up": {
      "message": {
        "type": "CycleFocusWorkspace",
        "content": "Next"
      }
    },
    "on_scroll_down": {
      "message": {
        "type": "CycleFocusWorkspace",
        "content": "Previous"
      }
    },
    "horizontal_scroll_threshold": 100,
    "on_scroll_left": {
      "message": {
        "type": "CycleStackIndex",
        "content": "Previous"
      }
    },
    "on_scroll_right": {
      "message": {
        "type": "CycleStackIndex",
        "content": "Next"
      }
    }
  },
```

CtByte commented:
@LGUG2Z Do you think I should just expose the `SocketMessage` enum, instead of making a new `MouseCommand`?

I just wanted to make it simple for users, so they do not need to remember to add `FocusMonitorAtCursor` and the `MouseFollowsFocus` (2 times!) when they set things up in the config. These "wrapper" messages could be there by default and you could disable them in the config perhaps 🤔

LGUG2Z commented:
```
    // The variable will be resolved
    //"on_primary_double_click": "komorebic replace-configuration $Env:USERPROFILE/komorebi.json",
    
    // This is the same as below but it feels slower
    // "on_primary_double_click": "komorebic toggle-pause",
```

For this we can probably spawn a shell instance which is persistent throughout the lifetime of the process like we do for `whkd`, because the startup time of both `powershell.exe` and `cmd.exe` is quite slow

LGUG2Z commented:
Relevant blurb from whkd: https://github.com/LGUG2Z/whkd/blob/b4fdfa911afc19e94279c58aae4aee2e77508173/src/main.rs#L135-L151

LGUG2Z commented:
Like I mentioned on the Discord it feels like there are some issues live-updating the config - for example, I initially started with just the "click" hooks, and then while it was running I added the "scroll" hooks, but I could not get the callbacks to run until I restarted the process 🤔 

I'm also still trying to get the horizontal scroll hooks working on an MX Master 3

CtByte commented:
> Like I mentioned on the Discord it feels like there are some issues live-updating the config - for example, I initially started with just the "click" hooks, and then while it was running I added the "scroll" hooks, but I could not get the callbacks to run until I restarted the process 🤔
> 
> I'm also still trying to get the horizontal scroll hooks working on an MX Master 3

I've added debug messages and fixed the issue with the hot-reload.

As for the vertical scroll, I think you should try my suggestions on [discord](https://discord.com/channels/898554690126630914/898556726108901386/1366298816483168299) as it works over here 😅 

Not sure when I have time for the persistent shell.

CtByte commented:
> ```
>     // The variable will be resolved
>     //"on_primary_double_click": "komorebic replace-configuration $Env:USERPROFILE/komorebi.json",
>     
>     // This is the same as below but it feels slower
>     // "on_primary_double_click": "komorebic toggle-pause",
> ```
> 
> For this we can probably spawn a shell instance which is persistent throughout the lifetime of the process like we do for `whkd`, because the startup time of both `powershell.exe` and `cmd.exe` is quite slow

I did what I can. I've decided to spawn a powershell.

It does work for me, please let me know if there is anything obvious I overlooked.

LGUG2Z commented:
```json
    "on_scroll_left": {
      "message": {
        "type": "CycleStackIndex",
        "content": "Previous"
      }
    },
    "on_scroll_right": {
      "message": {
        "type": "CycleStackIndex",
        "content": "Next"
      }
    }
```

These two are still not working for me even on apps which don't have a horizontal scroll behaviour, but I don't think this needs to be a blocker for now - I think we'll get this merged in and shipped with the next release, and maybe just hide the horizontal scroll options from the schema docs until there is a clear set of instructions we can add to the docs in order to get this working consistently for everybody
